Fix CCI report enable mechanism

markfoodyburton opened the pull request.

eyck commented on report/include/scp/report.h:
```diff
@@ -352,10 +352,11 @@ struct ScLogger {
-                                           sc_core::SC_MEDIUM) \
-        .type(__VA_ARGS__)                                     \
+#define SCP_WARN(...)                                             \
+    if (::scp::get_log_verbosity(__VA_ARGS__) >= sc_core::SC_LOW) \
```
Why should this check for SC_LOW be added? If you want to switch off Warnings just set the log level to ERROR and you are done. Aside of this warnings will not be counted if verbosity is set to SC_LOW although even in the SystemC implementation of reporting verbosity will only taken into account if severity is SC_INFO

markfoodyburton replied:
I think the issue is that there was no way to distinguish between warnings and errors. (In the case somebody wants errors, but not warnings).

eyck replied:
As earlier mentioned: you set the reporting level to log::ERROR and you are done. Yes, the filtering is done in the report handler itself so it goes thru the entire stack. But if you have that many warnings that this makes a difference you are doomed anyways ;-)

markfoodyburton replied:
Thats not what we found.
The use case is that we would like to be able to choose (freely) the level of reporting for each 'unit' (sc_module, or named thing).
Hence, to be able to control this with CCI, the 'global' level has to be set to print _all_ messages. The back-end has to print everything.
Then the CCI value will be used to set the actual level for each unit. This patch adds a hierarchy to this, which makes this somewhat more convenient to set - for instance by setting a 'global' CCI value.
But then the CCI value must be able to select which level you want, which includes e.g. printing only errors. That didn't seem possible before.

eyck replied:
Good point. I could not imagine that you want to switch of warnings on a component basis since in our projects we treat warnings as errors.

markfoodyburton replied:
(I'm not saying treating warnings as errors is a bad idea - but :-) )

eyck commented on report/src/report.cpp:
```diff
@@ -510,24 +510,51 @@ auto scp::get_log_verbosity(char const* str) -> sc_core::sc_verbosity {
+                    if (pos == std::string::npos) {
+                        current_name = "";
+                    } else {
+                        auto parent_level = current_name.substr(0, pos);
```
parent_level is redundant

eyck commented on report/src/report.cpp:
```diff
@@ -510,24 +510,51 @@ auto scp::get_log_verbosity(char const* str) -> sc_core::sc_verbosity {
-            return ret;
+        string current_name = std::string(str);
+        while (true) {
+            string param_name;
```
Should be declared outside the loop to avoid redundant constructor calls

markfoodyburton replied:
Doubt this makes much difference, but I think it would be better to construct it with an inline if, rather than constructing, deleting, copying it....

eyck commented on report/src/report.cpp:
```diff
@@ -510,24 +510,48 @@ auto scp::get_log_verbosity(char const* str) -> sc_core::sc_verbosity {
+                                    : current_name +
+                                          "." SCP_LOG_LEVEL_PARAM_NAME;
+
+            auto h = cci::cci_get_broker().get_param_handle(param_name);
```
using this directly prevents the use of the SCP* macros from sc_main. You will get
`Error: /Accellera/CCI/cci_get_broker: You may not request a broker from outside the SystemC hierarchy (use cci::cci_get_global_broker)` at runtime.

It might be better to use
```
auto broker = sc_core::sc_get_current_object()? cci::cci_get_broker(): cci::cci_get_global_broker(originator);
auto h = broker.get_param_handle(param_name);
```

and adding 
```
#ifdef HAS_CCI
thread_local cci::cci_originator originator("reporting");
#endif
```` 
in the anonymous namespace.

markfoodyburton replied:
Good catch and NICE idea !

eyck replied:
Well, I'm already bitten by that....
It should be changed in line 528 as well

markfoodyburton commented on report/src/report.cpp:
```diff
@@ -510,24 +510,48 @@ auto scp::get_log_verbosity(char const* str) -> sc_core::sc_verbosity {
+                lut[k] = ret;
+                return ret;
+            } else {
+                auto val = cci::cci_get_broker().get_preset_cci_value(
```
yes

markfoodyburton commented:
Qualcomm contribution needs to be changed to using the standard Apache header and add to the notice file

markfoodyburton commented:
@eyck I have updated this pull request to handle being used along with a CCI broker, either being constructed before or after the initialisation of the report library. I've also fixed a (self inflicted) issue with multi-threading.

I'll check with @barnasc but I would rather leave the license headers as they are for now.

I'll wait for your "OK"s before merging

eyck commented on report/src/report.cpp:
```diff
@@ -67,7 +73,7 @@ struct ExtLogConfig : public scp::LogConfig {
 };
 
-thread_local ExtLogConfig log_cfg;
+ExtLogConfig log_cfg;
```
By removing the thread_local here the whole thing is not thread safe anymore since the (re-)init_logging functions modify it without a guard.

markfoodyburton replied:
But leaving it as thread_local makes the whole thing not work in a multi-threaded environment. So what you're saying is that we should put a guard in init, in case 2 threads call the init?.

markfoodyburton replied:
(Actually that lock already exists within configure_logging - maybe it could be pulled out, but if you have two things 'fighting' over what sort of logging you want, and they have different setups, you're going to get non-deterministic behaviour. I dont think a lock is actually going to help much????)

eyck replied:
Actually the orignal implementation (with thread_local) is thread save and used in such an environment where 2 SystemC kernels run in 2 threads each having their own sc_main with the initialization of the logging infrastructure. The only downside is that you do not have a ExtLogConfig object where all threads share the instance. This boils down to one kernel with many threads vs. many kernels with one thread each.

But to make it thread save all the (re-)init_logging(..) and set_logging_level(..) would need to lock a mutex to protect the log_cfg instance

markfoodyburton replied:
> Actually the orignal implementation (with thread_local) is thread save and used in such an environment where 2 SystemC kernels run in 2 threads each having their own sc_main with the initialization of the logging infrastructure. The only downside is that you do not have a ExtLogConfig object where all threads share the instance. This boils down to one kernel with many threads vs. many kernels with one thread each.
> 
> But to make it thread save all the (re-)init_logging(..) and set_logging_level(..) would need to lock a mutex to protect the log_cfg instance

Ahh, I see - yes - maybe this should be an option - it's kind of a choice about the use case.
a #def ?

eyck replied:
Fine with me...

markfoodyburton replied:
> I was thinking about it. How would you name it?

How about DISABLE_REPORT_THREAD_LOCAL

markfoodyburton replied:
I'll roll something

markfoodyburton replied:
how does that look?

markfoodyburton commented:
@barnasc @eyck Are we OK to merge this now?

eyck commented on report/src/report.cpp:
```diff
@@ -190,7 +190,7 @@ auto compose_message(const sc_report& rep, const scp::LogConfig& cfg)
         if (*rep.get_msg())
             os << rep.get_msg();
-        if (rep.get_severity() > SC_INFO) {
+        if (rep.get_severity() >= cfg.file_info_from) {
```
Using >= here and the default for file_info_from being sc_core::SC_INFO does not resemble the original behavior. Is this intended?

markfoodyburton replied:
That's an error - I prefer the >= I think it makes more sense, but I intended to leave the thing as it was as a default. I'll fix.

markfoodyburton replied:
Fixed I believe

eyck replied:
Yes, indeed.

eyck commented:
> @barnasc @eyck Are we OK to merge this now?

Fine from my side.

barnasc commented:
> @barnasc @eyck Are we OK to merge this now?

If I recall correctly, we agreed to have some light-weight documentation (e.g. README file or the like) to document the use of the switches such as `HAS_CCI` and `DISABLE_REPORT_THREAD_LOCAL`. Was that added somewhere?

markfoodyburton commented:
> > @barnasc @eyck Are we OK to merge this now?
> 
> If I recall correctly, we agreed to have some light-weight documentation (e.g. README file or the like) to document the use of the switches such as `HAS_CCI` and `DISABLE_REPORT_THREAD_LOCAL`. Was that added somewhere?

We don't have such text yes - I'll try and work on some. @barnasc is there a 'wiki' flow for that?

barnasc commented:
> I'll try and work on some. @barnasc is there a 'wiki' flow for that?

We do not have wiki based documentation flow yet. My expectation was more that we document these switches in a plain text in a README file at the root of the utility tree.

markfoodyburton commented:
Still lacking documentation, but thread local issue now has a fix.

markfoodyburton commented:
BTW, the 'documentation' we have (which is very limited indeed) is already on the PR.

eyck commented on report/include/scp/report.h:
```diff
@@ -55,6 +74,28 @@ enum class log {
     DBGTRACE = TRACEALL
 };
+
+static const std::array<sc_core::sc_severity, 8> severity = {
```
This leads to code bloat as **any** translation unit instantiates this array. Declare it extern and  change the definition in report.cpp by removing the static modifier

eyck commented on report/include/scp/report.h:
```diff
@@ -55,6 +74,28 @@ enum class log {
+    sc_core::SC_INFO,    // scp::log::TRACE
+    sc_core::SC_INFO     // scp::log::TRACEALL
+};
+static const std::array<sc_core::sc_verbosity, 8> verbosity = {
```
Same here as with `severity`

markfoodyburton replied:
I think while I was playing around, I moved these un-necissarily I'll put them back to where they were.... (in .cpp) No need for them to be exposed in the .h

eyck commented:
Actually I would not cram the fmt support into the exisiting report.h as the integration does not bring reasonable value. As far as I can see the advantage is that instead of 
```
SCCINFO()<<fmt::format("FMT: {}", "printout")
```
I write 
```
SCCINFO()("FMT: {}", "printout")
```
But I loose the choice to use fmt when desirable; with the latest change I'm even _forced_ to use fmt. Moreover one does not gain anything as a ostringstream is going to be created anyways. With just adding a macro F you would get the same result but be more flexible:
```
#define F(...) <<fmt::format(__VA_ARGS__)
...
SCCINFO()F("FMT: {}", "printout")
````
which is fairly close to the current implementation but leaves the user the choice on a per statement basis.

A proper fmt/std::format macro support exclusively focusing on the use of fmt should setup the macros differently in a way that not std::ostringstream gets created rather the

markfoodyburton commented:
Hi Eyck, I’ll answer in detail here 😊

On 15/01/2023, 18:30, "Eyck Jentzsch" ***@***.***> wrote:


WARNING: This email originated from outside of Qualcomm. Please be wary of any links or attachments, and do not enable macros.

Actually I would not cram the fmt support into the exisiting report.h as the integration does not bring reasonable value. As far as I can see the advantage is that instead of

SCCINFO()<<fmt::format("FMT: {}", "printout")

I write

SCCINFO()("FMT: {}", "printout")



Exactly - and those two things are basically the same.



But I loose the choice to use fmt when desirable; with the latest change I'm even forced to use fmt.

There is no ‘forcing’ here - you can STILL do SCP_INFO() << “What you want “<< fmt::format(“fill your boots”);

Moreover one does not gain anything as a ostringstream is going to be created anyways. With just adding a macro F you would get the same result but be more flexible:

#define F(...) <<fmt::format(__VA_ARGS__)

...

SCCINFO()F("FMT: {}", "printout")



Why is the ‘more’ flexible - it’s the same? (with an extra ‘F’)



which is fairly close to the current implementation but leaves the user the choice on a per statement basis.

The user _CAN_ choose on a per statement basis… see the patch (and the example - which has a mix).

A proper fmt/std::format macro support exclusively focusing on the use of fmt should setup the macros differently in a way that not std::ostringstream gets created rather the

That’s basically what I’ve done



Cheers
Mark.



—
Reply to this email directly, view it on GitHub<https://github.com/accellera-official/systemc-common-practices/pull/16#issuecomment-1383207185>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AC3M3HJHI2IN2QQHNCKD5PDWSQXZ7ANCNFSM6AAAAAARDDF55Q>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

markfoodyburton commented:
> Actually I would not cram the fmt support into the exisiting report.h as the integration does not bring reasonable value. As far as I can see the advantage is that instead of
> 
> ```
> SCCINFO()<<fmt::format("FMT: {}", "printout")
> ```
> 
> I write
> 
> ```
> SCCINFO()("FMT: {}", "printout")
> ```
> 
> But I loose the choice to use fmt when desirable; with the latest change I'm even _forced_ to use fmt. Moreover one does not gain anything as a ostringstream is going to be created anyways. With just adding a macro F you would get the same result but be more flexible:
> 
> ```
> #define F(...) <<fmt::format(__VA_ARGS__)
> ...
> SCCINFO()F("FMT: {}", "printout")
> ```
> 
> which is fairly close to the current implementation but leaves the user the choice on a per statement basis.
> 
> A proper fmt/std::format macro support exclusively focusing on the use of fmt should setup the macros differently in a way that not std::ostringstream gets created rather the

I'm not sure this is a valid criticism, your suggestion (F(...)) is basically what I've done, but without the "F" and with no ambiguity.

markfoodyburton commented:
Hi Marrk,

actually the user does not have a choice on a per-statement basis. There 
are 2 cases:

 1. If you set FMT_SHARED then _SCP_FMT_EMPTY_STR is defined as a call
    to fmt::format and every instantiation of the macro SCP_LOG uses it.
    So no choice to use the macro without fmt...
 2. If FMT_SHARED is not set, _SCP_FMT_EMPTY_STR  is set to "Please add
    FMT library for FMT support." which is then added to each log
    output. So no choice either...

BR

-Eyck

MINRES Technologies GmbH
Firmensitz: 	Keltenhof 2, 85579 Neubiberg
Registergericht: 	Amtsgericht München, HRB 202036
Geschäftsführer: 	Eyck Jentzsch



Am 16.01.2023 um 08:57 schrieb Mark Burton:
>
> Hi Eyck, I’ll answer in detail here 😊
>
> On 15/01/2023, 18:30, "Eyck Jentzsch" ***@***.***> wrote:
>
> *WARNING:*This email originated from outside of Qualcomm. Please be 
> wary of any links or attachments, and do not enable macros.
>
> Actually I would not cram the fmt support into the exisiting report.h 
> as the integration does not bring reasonable value. As far as I can 
> see the advantage is that instead of
>
> |SCCINFO()<<fmt::format("FMT: {}", "printout")|
>
> I write
>
> |SCCINFO()("FMT: {}", "printout")|
> ||
> |Exactly - and those two things are basically the same.|
> ||
>
> But I loose the choice to use fmt when desirable; with the latest 
> change I'm even /forced/ to use fmt.
>
> There is no ‘forcing’ here - you can STILL do SCP_INFO() << “What you 
> want “<< fmt::format(“fill your boots”);
>
> Moreover one does not gain anything as a ostringstream is going to be 
> created anyways. With just adding a macro F you would get the same 
> result but be more flexible:
>
> |#define F(...) <<fmt::format(__VA_ARGS__)|
> |...|
> |SCCINFO()F("FMT: {}", "printout")|
> ||
> |Why is the ‘more’ flexible - it’s the same? (with an extra ‘F’)|
> ||
>
> which is fairly close to the current implementation but leaves the 
> user the choice on a per statement basis.
>
> The user _/CAN/_ choose on a per statement basis… see the patch (and 
> the example - which has a mix).
>
> A proper fmt/std::format macro support exclusively focusing on the use 
> of fmt should setup the macros differently in a way that not 
> std::ostringstream gets created rather the
>
> That’s basically what I’ve done
>
> Cheers
> Mark.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/accellera-official/systemc-common-practices/pull/16#issuecomment-1383207185>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AC3M3HJHI2IN2QQHNCKD5PDWSQXZ7ANCNFSM6AAAAAARDDF55Q>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

markfoodyburton commented:
On 16/01/2023, 09:29, "Eyck Jentzsch" ***@***.***> wrote:


Hi Marrk,

actually the user does not have a choice on a per-statement basis. There are 2 cases:
1.       If you set FMT_SHARED then  _SCP_FMT_EMPTY_STR is defined as a call to fmt::format and every instantiation of the macro SCP_LOG uses it. So no choice to use the macro without fmt...
No??????
I mean I could have a bug, but this seems to work:
        SCP_INFO() << "Uncached version empty";
        SCP_INFO(())("FMT String : Cached version default");
        SCP_INFO(SCMOD) << "UnCached version feature using SCMOD macro";
        SCP_INFO((m_my_logger)) << "Cached version using (m_my_logger)";
        SCP_INFO((D)) << "Cached version with D";
First one is “with nothing special at all”
Second one is with FMT formatting, and a ‘local’ (default) cache
3rd one uses the global LUT, with the SystemC name
4th and 5th ones uses a local named cache (the ‘D’ is because I was making sure we didn’t get a name conflict with other variables)

2.       If FMT_SHARED is not set, _SCP_FMT_EMPTY_STR  is set to "Please add FMT library for FMT support." which is then added to each log output. So no choice either...
If you build without FMT (there’s an option) then you get this IF you use (or a model you pull in uses) FMT. The plan being to ‘encourage’ its use, but not to ‘force’ it (at least it compiles/runs without).
Cheers
Mark.


BR

-Eyck
MINRES Technologies GmbH
Firmensitz:
Keltenhof 2, 85579 Neubiberg
Registergericht:
Amtsgericht München, HRB 202036
Geschäftsführer:
Eyck Jentzsch

Am 16.01.2023 um 08:57 schrieb Mark Burton:
Hi Eyck, I’ll answer in detail here 😊

On 15/01/2023, 18:30, "Eyck Jentzsch" ***@***.***>***@***.***> wrote:


WARNING: This email originated from outside of Qualcomm. Please be wary of any links or attachments, and do not enable macros.

Actually I would not cram the fmt support into the exisiting report.h as the integration does not bring reasonable value. As far as I can see the advantage is that instead of

SCCINFO()<<fmt::format("FMT: {}", "printout")

I write

SCCINFO()("FMT: {}", "printout")



Exactly - and those two things are basically the same.



But I loose the choice to use fmt when desirable; with the latest change I'm even forced to use fmt.

There is no ‘forcing’ here - you can STILL do SCP_INFO() << “What you want “<< fmt::format(“fill your boots”);

Moreover one does not gain anything as a ostringstream is going to be created anyways. With just adding a macro F you would get the same result but be more flexible:

#define F(...) <<fmt::format(__VA_ARGS__)

...

SCCINFO()F("FMT: {}", "printout")



Why is the ‘more’ flexible - it’s the same? (with an extra ‘F’)



which is fairly close to the current implementation but leaves the user the choice on a per statement basis.

The user _CAN_ choose on a per statement basis… see the patch (and the example - which has a mix).

A proper fmt/std::format macro support exclusively focusing on the use of fmt should setup the macros differently in a way that not std::ostringstream gets created rather the

That’s basically what I’ve done



Cheers
Mark.



—
Reply to this email directly, view it on GitHub<https://github.com/accellera-official/systemc-common-practices/pull/16#issuecomment-1383207185>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AC3M3HJHI2IN2QQHNCKD5PDWSQXZ7ANCNFSM6AAAAAARDDF55Q>.
You are receiving this because you authored the thread.Message ID: ***@***.***>***@***.***>

eyck commented:
I tried the pr and failed:
```
CMake Error at build/_deps/systemclanguage-build/SystemCLanguageConfig.cmake:44 (include):
  The file

    /home/eyck/git/systemc-common-practices/build/_deps/systemclanguage-build/SystemCLanguageTargets.cmake

  was generated by the export() command.  It may not be used as the argument
  to the include() command.  Use ALIAS targets instead to refer to targets by
  alternative names.

Call Stack (most recent call first):
  build/_deps/systemccci-src/CMakeLists.txt:365 (find_package)
```

Any idea?

eyck commented:
This is `cmake version 3.22.1` on Ubuntu 22.04.1 LTS ...

markfoodyburton commented:
Sort out SystemC tag
Explain in top level how to add a package lock? (and explain that this builds of main).


We probably need something like below (to answer @eyck)

  addpackage(
    NAME
    RapidJSON
    GIT_REPOSITORY
    https://github.com/Tencent/rapidjson
    GIT_TAG
    e0f68a435610e70ab5af44fc6a90523d69b210b3
    GIT_SHALLOW
    FALSE
    OPTIONS
    "RAPIDJSON_BUILD_TESTS OFF"
    "RAPIDJSON_BUILD_DOC OFF"
    "RAPIDJSON_BUILD_EXAMPLES OFF")

  set(RapidJSON_DIR "${RapidJSON_BINARY_DIR}")

eyck commented:
I still have an issue building the scp with the pr. As reference I'll attach a complete [build.log](https://github.com/accellera-official/systemc-common-practices/files/10571790/build.log) output 
Looking into build/_deps I do not see a checkout of RapidJSON, so this seem to not work. Having no experience with CPM I'm stuck...

eyck commented:
Moving the cpmaddpackage of RapidJSON before the one of SystemC CCI solved one problem but lets me stumble into the next one:
```
-- CPM: report: adding package SystemCCCI@ (master)
-- Checking OS and processor type.
CMake Error at _build/Default/_deps/systemclanguage-build/SystemCLanguageConfig.cmake:44 (include):
-- Using SystemC 2.3.4.20221128 (/home/eyck/git/systemc-common-practices/_build/Default/_deps/systemclanguage-build)
  The file

    /home/eyck/git/systemc-common-practices/_build/Default/_deps/systemclanguage-build/SystemCLanguageTargets.cmake

  was generated by the export() command.  It may not be used as the argument
  to the include() command.  Use ALIAS targets instead to refer to targets by
  alternative names.

Call Stack (most recent call first):
  _build/Default/_deps/systemccci-src/CMakeLists.txt:365 (find_package)
```

eyck commented:
I pass now the configuration step but I get a compile error (gcc 11 on ubuntu 22.4):
```
In file included from /home/eyck/git/systemc-common-practices/report/src/report.cpp:23:
/home/eyck/git/systemc-common-practices/report/include/scp/report.h:434:24: error: ‘sc_core::sc_module’ is not a class member
  434 |         -> decltype(p->sc_core::sc_module::name(), std::true_type());
      |                        ^~~~~~~
/home/eyck/git/systemc-common-practices/report/include/scp/report.h:434:33: error: ‘sc_core::sc_module’ has not been declared
  434 |         -> decltype(p->sc_core::sc_module::name(), std::true_type());
      |                                 ^~~~~~~~~
/home/eyck/git/systemc-common-practices/report/include/scp/report.h: In member function ‘std::enable_if_t<has_method<TYPE>, const char*> scp::call_sc_name_fn::operator()(TYPE*) const’:
/home/eyck/git/systemc-common-practices/report/include/scp/report.h:446:28: error: ‘sc_core::sc_module’ has not been declared
  446 |         return p->sc_core::sc_module::name();
      |                            ^~~~~~~~~
gmake[2]: *** [_deps/report-build/CMakeFiles/reporting.dir/build.make:76: _deps/report-build/CMakeFiles/reporting.dir/src/report.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:2670: _deps/report-build/CMakeFiles/reporting.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

eyck commented:
After fiddling a bit I got to the root cause. There is a `#include <sysc/kernel/sc_module.h>`missing...

markfoodyburton commented:
LOL
Which is why it always works for me I guess…. Where are you suggesting we put it?

Cheers
Mark.


> On 27 Feb 2023, at 16:24, Eyck Jentzsch ***@***.***> wrote:
> 
> WARNING: This email originated from outside of Qualcomm. Please be wary of any links or attachments, and do not enable macros.
> 
> 
> After fiddling a bit I got to the root cause. There is a #include <sysc/kernel/sc_module.h>missing...
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/accellera-official/systemc-common-practices/pull/16#issuecomment-1446532299>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AC3M3HNXFLZKSV7HU6KNOUDWZTBKDANCNFSM6AAAAAARDDF55Q>.
> You are receiving this because you authored the thread.
>

eyck commented:
Looks goog to me
